### PR TITLE
Unify StackManager Delete funcs

### LINKED ISSUE #5451: [Task] Unify the different `Delete` methods
### Description

See comment [here](https://github.com/weaveworks/eksctl/issues/4655#issuecomment-1041834989)

- (7~ methods) Tasks for creating/deleting stacks (returning a `*tasks.TaskTree`) e.g.`NewTasksToCreateIAMServiceAccounts`, `NewTasksToDeleteIAMServiceAccounts`, `DeleteTasksForDeprecatedStacks`
- (4~ methods) Deleting stacks directly (not via task), e.g. `DeleteStackSync(s *Stack) error`, `DeleteStackBySpec(s *Stack) (*Stack, error)`

We need to 
- Unify the different `Delete` methods (@skarlso might of tried this recently?) into 1 func that has a `DeleteOpts` that allow you to customise how the deletion works (via spec/name, sync/async)

### PR DESCRIPTION
**Pull request by vflaux**
### Description

Unifies StackManager funcs `DeleteStackBySpec`, `DeleteStackBySpecSync` & `DeleteStackSync` into one `DeleteStack` func (#5451).
```go
type DeleteStackOptions struct {
	Stack *Stack
	Wait  bool
	ErrCh chan error
}

// DeleteStack sends a request to delete the stack.
// If Wait is true, it waits until status is DELETE_COMPLETE;
// If ErrCh is given, any errors will be written to it, assume completion when nil is written,
// do not expect more than one error value on the channel, it's closed immediately after it is written to.
// Wait and ErrCh are mutually exclusive
DeleteStack(ctx context.Context, options DeleteStackOptions) error
```

Changes `taskWithStackSpec` & `asynctaskWithStackSpec` into `deleteStackTask` with a wait option.
Also add a context parameter to `NewTasksToDeleteNodeGroups` func.

### Checklist
- [x] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [ ] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [x] Refactored something and made the world a better place :star2:

### CODE REVIEW

File: pkg/cfn/manager/api.go [hunk 1]
```diff
@@ -552,8 +552,34 @@ func defaultStackStatusFilter() []types.StackStatus {
 	return allNonDeletedStackStatuses()
 }
 
-// DeleteStackBySpec sends a request to delete the stack
-func (c *StackCollection) DeleteStackBySpec(ctx context.Context, s *Stack) (*Stack, error) {
+// DeleteStack sends a request to delete the stack.
+// If Wait is true, it waits until status is DELETE_COMPLETE;
+// If ErrCh is given, any errors will be written to it, assume completion when nil is written,
+// do not expect more than one error value on the channel, it's closed immediately after it is written to.
+// Wait and ErrCh are mutually exclusive
```
Comment (TiberiuGC) [inline pkg/cfn/manager/api.go:559]:
```suggestion
// ErrCh is essentially the mechanism used to wait for completion asynchronously,
// for the circumstances when this method is invoked as part of a _deleteStackTask_ .
// Wait and ErrCh are mutually exclusive.
```

File: pkg/cfn/manager/api.go [hunk 2]
```diff
@@ -552,8 +552,34 @@ func defaultStackStatusFilter() []types.StackStatus {
 	return allNonDeletedStackStatuses()
 }
 
-// DeleteStackBySpec sends a request to delete the stack
-func (c *StackCollection) DeleteStackBySpec(ctx context.Context, s *Stack) (*Stack, error) {
+// DeleteStack sends a request to delete the stack.
+// If Wait is true, it waits until status is DELETE_COMPLETE;
```
Comment (TiberiuGC) [inline pkg/cfn/manager/api.go:556]:
```suggestion
// If Wait is true, it waits synchronously until status is DELETE_COMPLETE;
```

File: pkg/cfn/manager/api.go [hunk 3]
```diff
@@ -552,8 +552,34 @@ func defaultStackStatusFilter() []types.StackStatus {
...
+// If ErrCh is given, any errors will be written to it, assume completion when nil is written,
+// do not expect more than one error value on the channel, it's closed immediately after it is written to.
+// Wait and ErrCh are mutually exclusive
+func (c *StackCollection) DeleteStack(ctx context.Context, options DeleteStackOptions) error {
+	if _, err := c.doDeleteStack(ctx, options.Stack); err != nil {
```
Comment (TiberiuGC) [inline pkg/cfn/manager/api.go:561]: I may be missing something, but I don't see much value in this function returning back the stack? Could we perhaps refactor here as-well such that it only returns an error?

File: pkg/cfn/manager/deprecated.go [hunk 4]
```diff
@@ -74,9 +74,10 @@ func (c *StackCollection) DeleteTasksForDeprecatedStacks(ctx context.Context) (*
 				if suffix == "-ControlPlane" && !cpStackFound {
 					taskTree.Append(deleteControlPlaneTask)
 				} else {
-					taskTree.Append(&taskWithStackSpec{
-						stack: s,
-						call:  c.DeleteStackBySpecSync,
+					taskTree.Append(&deleteStackTask{
+						stack:           s,
+						stackCollection: c,
+						ctx:             ctx,
```
Comment (TiberiuGC) [inline pkg/cfn/manager/deprecated.go:80]: Shouldn't we also set `wait: true` here?

**Review by TiberiuGC — CHANGES_REQUESTED** [review]

**Comment by TiberiuGC** [comment]
@vflaux - beautiful piece of refactoring, thanks a lot for opening this PR ✨ !! From a code's logic perspective, LGTM. I've generally left minor stylistic comments.

File: pkg/cfn/manager/api.go (hunk 3 above)
Comment (vflaux) [reply pkg/cfn/manager/api.go]: As `doDeleteStack()` is only used here, there is indeed no point in returnning the stack.
I made the change on `doDeleteStack()`.

File: pkg/cfn/manager/deprecated.go (hunk 4 above)
Comment (vflaux) [reply pkg/cfn/manager/deprecated.go:80]: I think you're right as the previous code was synchronous.
I added it.

**Review by TiberiuGC — APPROVED** [review]

**Comment by vflaux** [comment]
I tested manually with a `delete nodegroup` on an unamanged nodegroup.

New:
```
$ ./eksct version
0.110.0-dev+22f59427b.2022-08-25T17:47:57Z
$ ./eksctl --cluster foo delete nodegroup --name ng-1
2022-08-25 17:49:22 [ℹ]  1 nodegroup (ng-1) was included (based on the include/exclude rules)
2022-08-25 17:49:28 [ℹ]  will drain 1 nodegroup(s) in cluster "foo"
2022-08-25 17:49:28 [ℹ]  starting parallel draining, max in-flight of 1
2022-08-25 17:49:28 [!]  no nodes found in nodegroup "ng-1" (label selector: "alpha.eksctl.io/nodegroup-name=ng-1")
2022-08-25 17:49:28 [ℹ]  will delete 1 nodegroups from cluster "foo"
2022-08-25 17:49:37 [ℹ]  1 task: { 1 task: { delete nodegroup "ng-1" [async] } }
2022-08-25 17:49:37 [ℹ]  will delete stack "eksctl-foo-nodegroup-ng-1"
2022-08-25 17:49:37 [ℹ]  will delete 1 nodegroups from auth ConfigMap in cluster "foo"
2022-08-25 17:49:37 [ℹ]  removing identity "arn:aws:iam::123456789012:role/eksctl-foo-NodeInstanceRole-ABCDEFGHIJKLM" from auth ConfigMap (username = "system:node:{{EC2PrivateDNSName}}", groups = ["system:bootstrappers" "system:nodes"])
2022-08-25 17:49:37 [✔]  deleted 1 nodegroup(s) from cluster "foo"
```

Current:
```
$ eksctl version
0.109.0
$ eksctl --cluster foo delete nodegroup --name ng-1
2022-08-25 18:09:49 [ℹ]  1 nodegroup (ng-1) was included (based on the include/exclude rules)
2022-08-25 18:09:53 [ℹ]  will drain 1 nodegroup(s) in cluster "foo"
2022-08-25 18:09:53 [ℹ]  starting parallel draining, max in-flight of 1
2022-08-25 18:09:53 [!]  no nodes found in nodegroup "ng-1" (label selector: "alpha.eksctl.io/nodegroup-name=ng-1")
2022-08-25 18:09:53 [ℹ]  will delete 1 nodegroups from cluster "foo"
2022-08-25 18:10:03 [ℹ]  1 task: { 1 task: { delete nodegroup "ng-1" [async] } }
2022-08-25 18:10:03 [ℹ]  will delete stack "eksctl-foo-nodegroup-ng-1"
2022-08-25 18:10:03 [ℹ]  will delete 1 nodegroups from auth ConfigMap in cluster "foo"
2022-08-25 18:10:03 [ℹ]  removing identity "arn:aws:iam::123456789012:role/eksctl-foo-NodeInstanceRole-ABCDEFGHIJKLM" from auth ConfigMap (username = "system:node:{{EC2PrivateDNSName}}", groups = ["system:bootstrappers" "system:nodes"])
2022-08-25 18:10:04 [✔]  deleted 1 nodegroup(s) from cluster "foo"
```

**Pull request closed without merging**